backends/ebpf/psa/ebpfPipeline.cpp

```diff
@@ -672,4 +672,220 @@ void TCEgressPipeline::emitTrafficManager(CodeBuilder* builder) {
+    builder->emitIndent();
+    builder->appendFormat("return %s", dropReturnCode());
+    builder->endOfStatement(true);
+    builder->blockEnd(true);
+
+    builder->newline();
+
+    // normal packet to port
+    varStr = Util::printf_format("%s.egress_port", control->inputStandardMetadata->name.name);
+    builder->target->emitTraceMessage(builder, "EgressTM: output packet to port %d", 1, varStr);
+    builder->emitIndent();
+    builder->appendFormat("return %s;", this->forwardReturnCode());
+    builder->newline();
+}
+
+void XDPEgressPipeline::emitCheckPacketMarkMetadata(CodeBuilder* builder) {
+    (void)builder;
+    // Global metadata in XDP is not preserved across ingress and egress so do not check packet
+    // mark. Anyway, in XDP egress can't be called without our ingress.
+}
+
+// =====================TCTrafficManagerForXDP=============================
+
+void TCTrafficManagerForXDP::emitGlobalMetadataInitializer(CodeBuilder* builder) {
+    EBPFPipeline::emitGlobalMetadataInitializer(builder);
+
+    // if Traffic Manager decided to pass packet to the kernel stack earlier, send it up immediately
+    builder->emitIndent();
+    builder->appendFormat("if (%s->pass_to_kernel == true) return %s;", compilerGlobalMetadata,
+                          progTarget->forwardReturnCode());
+    builder->newline();
+
+    // Mark packet for egress processing
+    builder->emitIndent();
+    builder->appendFormat("%s->mark = %u", compilerGlobalMetadata, packetMark);
+    builder->endOfStatement(true);
+}
+
+void TCTrafficManagerForXDP::emit(CodeBuilder* builder) {
+    cstring msgStr;
+    progTarget->emitCodeSection(builder, sectionName);
+    builder->emitIndent();
+    progTarget->emitMain(builder, functionName, model.CPacketName.str());
+    builder->spc();
+    builder->blockStart();
+    emitGlobalMetadataInitializer(builder);
+    emitLocalVariables(builder);
+
+    if (options.xdp2tcMode == XDP2TC_CPUMAP) {
+        emitReadXDP2TCMetadataFromCPUMAP(builder);
+    } else if (options.xdp2tcMode == XDP2TC_HEAD) {
+        emitReadXDP2TCMetadataFromHead(builder);
+    }
+
+    msgStr = Util::printf_format("%s deparser: packet deparsing started", sectionName);
+    builder->target->emitTraceMessage(builder, msgStr.c_str());
+    builder->emitIndent();
+    deparser->emit(builder);
+    msgStr = Util::printf_format("%s deparser: packet deparsing finished", sectionName);
+    builder->target->emitTraceMessage(builder, msgStr.c_str());
+    this->emitTrafficManager(builder);
+
+    builder->blockEnd(true);
+}
+
+void TCTrafficManagerForXDP::emitReadXDP2TCMetadataFromHead(CodeBuilder* builder) {
+    builder->emitIndent();
+    builder->append(
+        "    void *data = (void *)(long)skb->data;\n"
+        "    void *data_end = (void *)(long)skb->data_end;\n"
+        "    if (((char *) data + 14 + sizeof(struct xdp2tc_metadata)) > (char *) data_end) {\n"
```

Comment:
14?

Reply:
The size of Ethernet header. It could be replaced by a proper `sizeof`.